### PR TITLE
feat(nms): Add Empty State component

### PR DESCRIPTION
**Pull request by HannaFar**
Signed-off-by: HannaFar <hannafarag159@gmail.com>

## Summary
Add empty state component.
Add empty state content for: 

Gateways
Capture d’écran 2022-03-29 à 18 22 52

eNodeBs: 
Capture d’écran 2022-03-31 à 18 28 41

Policies: 
Capture d’écran 2022-03-31 à 18 29 00

APNs: 
Capture d’écran 2022-03-31 à 18 29 10

Subscribers: 
Capture d’écran 2022-03-31 à 18 28 50

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: nms/packages/magmalte/app/views/equipment/EquipmentGateway.js [hunk 1]
```diff
@@ -21,35 +22,57 @@ import type {
 
 import ActionTable from '../../components/ActionTable';
 import AutorefreshCheckbox from '../../components/AutorefreshCheckbox';
+// import Button from '@material-ui/core/Button';
+// import Card from '@material-ui/core/Card';
+// import CardActions from '@material-ui/core/CardActions';
+// import CardContent from '@material-ui/core/CardContent';
+
+// import CardHeader from '@material-ui/core/CardHeader';
```
Comment (andreilee) [inline nms/packages/magmalte/app/views/equipment/EquipmentGateway.js]: Clean these up

File: nms/packages/magmalte/app/views/equipment/EquipmentGateway.js [hunk 2]
```diff
@@ -92,32 +123,162 @@ const useStyles = makeStyles(theme => ({
...
+            <EmptyState
+              title={'Set up a Gateway'}
+              customIntructions={<GatewayEmptyStateList />}
+              instructions={''}
+              buttonText={'Add Gateway'}
+              // onClick={() => }
+              linkText={'Learn more about Access Gateway Configuration'}
+              link={
+                'https://docs.magmacore.org/docs/next/lte/deploy_config_agw'
+              }
```
Comment (andreilee) [inline nms/packages/magmalte/app/views/equipment/EquipmentGateway.js]: Why is this specified both here and in 238-240?

**Review by andreilee — CHANGES_REQUESTED** [review]

File: nms/packages/magmalte/app/views/equipment/EquipmentGateway.js (hunk 2 above)
Comment (HannaFar) [reply nms/packages/magmalte/app/views/equipment/EquipmentGateway.js]: It should not sorry!

File: nms/packages/magmalte/app/components/EmptyState.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  title: string,
+  customIntructions?: React.Node,
+  instructions: string,
+  overviewTitle: string,
+  overviewDescription: string,
+  cardActions?: CardActionsType,
+};
+
+/**
+ * Empty state instructions. Display 2 card with instructions when there is no data to display.
```
Comment (andreilee) [inline nms/packages/magmalte/app/components/EmptyState.js]: Mind the lints

File: nms/packages/magmalte/app/views/equipment/EquipmentEnodeb.js [hunk 4]
```diff
@@ -80,26 +104,129 @@ const useStyles = makeStyles(theme => ({
...
 
+  return (
+    <Grid
+      className={classes.customIntructions}
+      spacing={3}
+      container
+      justify="space-between">
+      <Grid item xs={6}>
+        <Card className={classes.instructions}>
+          {/* <Text></Text> */}
```
Comment (andreilee) [inline nms/packages/magmalte/app/views/equipment/EquipmentEnodeb.js]: remove

File: nms/packages/magmalte/app/views/equipment/EquipmentGateway.js [hunk 5]
```diff
@@ -92,32 +119,162 @@ const useStyles = makeStyles(theme => ({
...
+              Learn more about Access Gateway Configuration
+            </Link>
+          </Grid>
+        </Grid>
+      </Grid>
+    </Grid>
+  );
+}
+
+function GatewayEmptyStateList() {
```
Comment (andreilee) [inline nms/packages/magmalte/app/views/equipment/EquipmentGateway.js]: Put these three new components into a new file

**Review by andreilee — APPROVED** [review]
Overall this looks good! Please fix the lints and other minor things.

File: nms/app/components/EmptyState.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright 2020 The Magma Authors.
```
Comment (thmsschmitt) [inline nms/app/components/EmptyState.js]: 2020 -> 2022

File: nms/app/components/EmptyState.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  const classes = useStyles();
+  const CustomIntructions = props.customIntructions;
+  return (
+    <Grid container className={classes.emptyState} spacing={6}>
+      <Grid item xs={8}>
+        <Card raised={false} className={classes.card}>
+          <CardHeader title={props.title} />
+          <CardContent>
+            <Text className={classes.instructions}>{props.instructions}</Text>
+            {(props.customIntructions ?? false) && CustomIntructions}
```
Comment (thmsschmitt) [inline nms/app/components/EmptyState.js]: That looks overly complicated.
Shouldn't `{props.customIntructions}` be sufficient?

File: nms/app/components/admin/Networks.js [hunk 8]
```diff
@@ -137,33 +138,49 @@ function Networks() {
...
-        </NestedRouteLink>
-      </div>
       {rows.length === 0 ? (
-        <div className={classes.noNetworks}>
-          <NoNetworksMessage>
-            You currently do not have any networks configured. Click "Add
-            Network" to create a new network
-          </NoNetworksMessage>
-        </div>
+        // <div className={classes.noNetworks}>
```
Comment (thmsschmitt) [inline nms/app/components/admin/Networks.js]: Please remove

File: nms/app/views/equipment/EquipmentEnodeb.js [hunk 9]
```diff
@@ -80,26 +102,131 @@ const useStyles = makeStyles(theme => ({
...
+            </ul>
+          </CardContent>
+          <CardActions disableSpacing={true}>
+            <Grid container direction="column" spacing={1}>
+              <Grid item xs={5}>
+                <Button
+                  variant="contained"
+                  color="primary"
+                  onClick={() => props.setOpen()}>
+                  {'Add enodeb'}
```
Comment (thmsschmitt) [inline nms/app/views/equipment/EquipmentEnodeb.js]: No need for the `{'`

File: nms/app/views/equipment/EquipmentEnodeb.js [hunk 10]
```diff
@@ -117,32 +244,37 @@ type EnodebRowType = {
...
-    networkId: networkId,
-    type: 'enodeb',
-    interval: REFRESH_INTERVAL,
-    refresh: refresh,
-    lastRefreshTime: lastRefreshTime,
-  });
-  const ctxValues = [...Object.values(ctx.state.enbInfo)];
+  const [lastFetchTime, setLastFetchTime] = useState(Date.now());
+  const [currRow, setCurrRow] = useState<EnodebRowType>({});
+  const [enbInfo, setEnbInfo] = useState(ctx.state?.enbInfo);
```
Comment (thmsschmitt) [inline nms/app/views/equipment/EquipmentEnodeb.js]: This again creates a local copy of the context. 
We should use `ctx.state?.enbInfo` directly below and should call `ctx.setState` in `fetchState` below.

File: nms/app/views/equipment/EquipmentEnodeb.js [hunk 11]
```diff
@@ -172,7 +304,7 @@ function EnodebTableRaw(props: WithAlert) {
       <CardTitleRow
         key="title"
         icon={SettingsInputAntennaIcon}
-        label={`Enodebs (${Object.keys(state?.enbInfo || {}).length})`}
+        label={`Enodebs (${Object.keys(ctx.state?.enbInfo || {}).length})`}
```
Comment (thmsschmitt) [inline nms/app/views/equipment/EquipmentEnodeb.js:293]: Here we are not using the local copy 🙈

File: nms/app/views/equipment/EquipmentGateway.js [hunk 12]
```diff
@@ -299,30 +462,38 @@ function GatewayStatusTable(props: WithAlert & {refresh: boolean}) {
...
-  }, [ctxValues.length]);
-
   const subscriberCtx = useContext(SubscriberContext);
   const gwSubscriberMap = subscriberCtx.gwSubscriberMap;
-
-  const lteGateways = state;
   const [currRow, setCurrRow] = useState<EquipmentGatewayRowType>({});
   const lteGatewayRows: Array<EquipmentGatewayRowType> = [];
+  const [lastFetchTime, setLastFetchTime] = useState(Date.now());
+  const [lteGateways, setLteGateways] = useState(gwCtx.state);
```
Comment (thmsschmitt) [inline nms/app/views/equipment/EquipmentGateway.js]: Same problem here

File: nms/app/views/equipment/EquipmentGateway.js [hunk 13]
```diff
@@ -299,30 +462,38 @@ function GatewayStatusTable(props: WithAlert & {refresh: boolean}) {
...
+  const REFRESH_INTERVAL = 30000;
+
+  useInterval(() => {
+    if (props.refresh) {
+      setLastFetchTime(Date.now());
+    }
+  }, REFRESH_INTERVAL);
+
+  useEffect(() => {
```
Comment (thmsschmitt) [inline nms/app/views/equipment/EquipmentGateway.js]: We do actually not need the `useEffect` and the  `lastFetchTime`.
We could just do something like:

```
useInterval(() => {
  FetchGateways({ networkId, enqueueSnackbar })
    .then(gateways => {
      if (gateways) {
        gwCtx.setSate(gateways);
      }
     });
  }, props.refresh ? REFRESH_INTERVAL: null);
  ```

File: nms/app/views/equipment/EquipmentEnodeb.js [hunk 14]
```diff
@@ -117,32 +244,37 @@ type EnodebRowType = {
...
+  const REFRESH_INTERVAL = 30000;
+  useInterval(() => {
+    if (refresh) {
+      setLastFetchTime(Date.now());
+    }
+  }, REFRESH_INTERVAL);
   useEffect(() => {
-    setLastRefreshTime(new Date().toLocaleString());
-  }, [ctxValues.length]);
+    const fetchState = async () => {
```
Comment (thmsschmitt) [inline nms/app/views/equipment/EquipmentEnodeb.js]: See below how this can be simplified

File: nms/app/views/equipment/EquipmentGateway.js [hunk 15]
```diff
@@ -91,32 +117,169 @@ const useStyles = makeStyles(theme => ({
...
+    <ul className={classes.bulletList}>
+      <li>
+        Create bootable USB with OS (Ubuntu).
+        <Link href={VIEW_DOCUMENTATION_LINK} target="_blank">
+          View documentation
+        </Link>
+      </li>
+      <li>Install Magma service</li>
+      <li>
+        Download <Link href={''}>rootca.pem</Link> and{' '}
```
Comment (thmsschmitt) [inline nms/app/views/equipment/EquipmentGateway.js:200]: Are those links supposed to be empty?

File: nms/app/views/equipment/EquipmentGateway.js [hunk 16]
```diff
@@ -299,30 +462,38 @@ function GatewayStatusTable(props: WithAlert & {refresh: boolean}) {
...
+      try {
+        const gateways = await FetchGateways({
+          networkId,
+          enqueueSnackbar,
+        });
+        if (gateways) {
+          setLteGateways(gateways);
+        }
+      } catch (e) {
+        enqueueSnackbar?.('failed fetching gateways', {
```
Comment (thmsschmitt) [inline nms/app/views/equipment/EquipmentGateway.js]: `FetchGateways` does actually already take care of the error handling

File: nms/app/views/equipment/EquipmentGateway.js [hunk 17]
```diff
@@ -299,30 +462,38 @@ function GatewayStatusTable(props: WithAlert & {refresh: boolean}) {
...
   const lteGatewayRows: Array<EquipmentGatewayRowType> = [];
+  const [lastFetchTime, setLastFetchTime] = useState(Date.now());
+  const [lteGateways, setLteGateways] = useState(gwCtx.state);
+  const REFRESH_INTERVAL = 30000;
+
+  useInterval(() => {
+    if (props.refresh) {
+      setLastFetchTime(Date.now());
+    }
+  }, REFRESH_INTERVAL);
```
Comment (thmsschmitt) [inline nms/app/views/equipment/EquipmentGateway.js]: If you pass NULL to `useInterval` the refreshing will be disabled, see below.

**Review by thmsschmitt — CHANGES_REQUESTED** [review]
Looks good, but the state management still needs some refinement

**Review by thmsschmitt — APPROVED** [review]
I think we can finally merge this, but please rebase it once more beforehand.

**Pull request merged**